x/fees/keeper/integration_test.go

```diff
@@ -0,0 +1,339 @@
+	)
+	totalAmount := sdk.NewCoin(claimsDenom, initAmount.MulRaw(int64(accountCount)))
+
+	var (
+		deployerKey      *ethsecp256k1.PrivKey
+		userKey          *ethsecp256k1.PrivKey
+		deployerAddress  sdk.AccAddress
+		userAddress      sdk.AccAddress
+		params           types.Params
+		contractAddress  common.Address
+		contractAddress1 common.Address
+		contractAddress2 common.Address
+	)
+
+	BeforeAll(func() {
+		s.SetupTest()
+
+		params = s.app.FeesKeeper.GetParams(s.ctx)
+		params.EnableFees = true
+		s.app.FeesKeeper.SetParams(s.ctx, params)
+
+		// mint coins for claiming and send them to the claims module
+		coins := sdk.NewCoins(totalAmount)
+		err := testutil.FundModuleAccount(s.app.BankKeeper, s.ctx, inflationtypes.ModuleName, coins)
+		s.Require().NoError(err)
+
+		// setup accounts
+		deployerKey, _ = ethsecp256k1.GenerateKey()
+		deployerAddress = getAddr(deployerKey)
+		testutil.FundAccount(s.app.BankKeeper, s.ctx, deployerAddress, initBalance)
+
+		userKey, _ = ethsecp256k1.GenerateKey()
+		userAddress = getAddr(userKey)
+		testutil.FundAccount(s.app.BankKeeper, s.ctx, userAddress, initBalance)
+		acc := s.app.AccountKeeper.NewAccountWithAddress(s.ctx, userAddress)
+		s.app.AccountKeeper.SetAccount(s.ctx, acc)
+		s.Commit()
+
+		// deploy contracts
+		contractAddress = deployContract(deployerKey)
+		contractAddress1 = deployContract(deployerKey)
+		contractAddress2 = deployContract(deployerKey)
+
+		// register a contract
+		registerDevFeeInfo(deployerKey, &contractAddress, 0)
+		fee, isRegistered := s.app.FeesKeeper.GetFeeInfo(s.ctx, contractAddress)
+		Expect(isRegistered).To(Equal(true))
+		Expect(fee.ContractAddress).To(Equal(contractAddress.Hex()))
+		Expect(fee.DeployerAddress).To(Equal(deployerAddress.String()))
+		Expect(fee.WithdrawAddress).To(Equal(""))
+		s.Commit()
+	})
+
+	Context("fee distribution is disabled", func() {
+		BeforeAll(func() {
+			params = s.app.FeesKeeper.GetParams(s.ctx)
+			params.EnableFees = false
+			s.app.FeesKeeper.SetParams(s.ctx, params)
+		})
+
+		It("we cannot register contracts for receiving tx fees", func() {
+			fromAddress := sdk.AccAddress(deployerKey.PubKey().Address().Bytes())
+			msg := types.NewMsgRegisterDevFeeInfo(contractAddress1, fromAddress, fromAddress, []uint64{1})
+
+			res := deliverTx(deployerKey, msg)
+			Expect(res.IsOK()).To(Equal(false), "registration should have failed")
+			s.Commit()
+
+			_, isRegistered := s.app.FeesKeeper.GetFeeInfo(s.ctx, contractAddress1)
+			Expect(isRegistered).To(Equal(false))
+		})
+
+		It("no tx fees go to developers", func() {
+			preBalance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, evmDenom)
+			gasPrice := big.NewInt(2000000000)
+			contractInteract(userKey, &contractAddress, gasPrice, nil, nil, nil)
+
+			balance := s.app.BankKeeper.GetBalance(s.ctx, deployerAddress, evmDenom)
+			Expect(balance).To(Equal(preBalance))
+		})
+	})
+
+	Context("fee distribution is enabled", func() {
```

Comment:
we also need to create cases for:

- params with 100% of fee assigned to validators: check that dev balance remains the same and only fee collector balance is updated
- params with 100% of fee assigned to developer: check that fee collector balance remains the same and only dev balance is updated
- factory generated contracts with legacy tx
- factory generated contracts with dynamic fee txs
- registration of dev fees: withdraw address is different than the deployer address
- update of withdrawal address: withdraw address is different than the deployer address
- canceled developer fees: check that no fees are distributed

Reply:
Done.